tensorflow_probability/python/experimental/mcmc/covariance_reducer_test.py

```diff
@@ -131,17 +146,88 @@ def test_nested_with_batching_and_chunking(self):
     cov_reducer = tfp.experimental.mcmc.CovarianceReducer(event_ndims=1)
     chain_state = ({'one': tf.ones((3, 4)), 'zero': tf.zeros((3, 4))},
                    {'two': tf.ones((3, 4)) * 2})
-    state = cov_reducer.initialize(chain_state)
+    fake_kr = FakeKernelResults(0, FakeInnerResults(0))
+    state = cov_reducer.initialize(chain_state, fake_kr)
     _, state = tf.while_loop(
         lambda i, _: i < 10,
-        lambda i, state: (i + 1, cov_reducer.one_step(chain_state, state, 0)),
+        lambda i, state: (i + 1, cov_reducer.one_step(
+            chain_state, state, fake_kr, 0)),
         (0., state)
     )
     final_cov = self.evaluate(cov_reducer.finalize(state))
     self.assertAllEqualNested(
         final_cov, ({'one': tf.zeros((3, 4, 4)), 'zero': tf.zeros((3, 4, 4))},
                     {'two': tf.zeros((3, 4, 4))}))
 
+  def test_manual_variance_transform_fn(self):
+    var_reducer = tfp.experimental.mcmc.VarianceReducer(
+        transform_fn=lambda _, kr: kr.inner_results.value)
+    fake_kr = FakeKernelResults(0., FakeInnerResults(
+        tf.zeros((2, 3))))
+    # chain state should be irrelevant
+    state = var_reducer.initialize(0., fake_kr)
+    for sample in range(5):
+      fake_kr = FakeKernelResults(
+          sample, FakeInnerResults(tf.ones((2, 3)) * sample))
+      state = var_reducer.one_step(sample, state, fake_kr)
+    final_mean, final_var = self.evaluate([
+        state.cov_state.mean,
+        var_reducer.finalize(state)])
+    self.assertEqual((2, 3), final_mean.shape)
+    self.assertAllEqual(np.ones((2, 3)) * 2, final_mean)
+    self.assertEqual((2, 3), final_var.shape)
+    self.assertAllEqual(np.ones((2, 3)) * 2, final_var)
+
+  def test_manual_covariance_transform_fn_with_random_states(self):
+    rng = test_util.test_np_rng()
+    x = rng.rand(100, 5, 2)
+    cov_reducer = tfp.experimental.mcmc.CovarianceReducer(
+        transform_fn=lambda _, kr: kr.inner_results.value)
+    fake_kr = FakeKernelResults(0., FakeInnerResults(
+        tf.zeros((5, 2))))
+    state = cov_reducer.initialize(0., fake_kr)
+    for sample in x:
+      fake_kr = FakeKernelResults(0., FakeInnerResults(sample))
+      state = cov_reducer.one_step(0., state, fake_kr)
+    final_mean, final_cov = self.evaluate([
+        state.cov_state.mean,
+        cov_reducer.finalize(state)])
+
+    # reshaping to be compatible with a check against numpy
+    x_reshaped = x.reshape(100, 10)
+    final_cov_reshaped = tf.reshape(final_cov, (10, 10))
+    self.assertEqual((5, 2), final_mean.shape)
+    self.assertAllClose(np.mean(x, axis=0), final_mean, rtol=1e-5)
+    self.assertEqual((5, 2, 5, 2), final_cov.shape)
+    self.assertAllClose(np.cov(x_reshaped.T, ddof=0),
+                        final_cov_reshaped,
+                        rtol=1e-5)
+
+  def test_latent_state_with_multiple_transform_fn(self):
```

Comment:
Also test that a transform_fn that returns a structure works as expected (in a tf.while_loop, no less).

If there is a reason to make it work differently than suggested in my previous comment, you will probably discover it.